Attachment: Use opaque background color to ensure readability

squelart opened the pull request:
#### 58e3ab08228ff163f000cba1a2786e9c9f7c9bbc
<pre>
Attachment: Use opaque background color to ensure readability
<a href="https://bugs.webkit.org/show_bug.cgi?id=267253">https://bugs.webkit.org/show_bug.cgi?id=267253</a>
<a href="https://rdar.apple.com/117749005">rdar://117749005</a>

Reviewed by Tim Nguyen.

The system colors used for the attachment background are very transparent,
which could render the attachment contents less readable if whatever is
under the attachment is of a similar color as the attachment text.
This issue can be prevented by making the background fully opaque.
The original transparent color has been moved to a new child div.

* LayoutTests/fast/attachment/mac/wide-attachment-image-controls-basic-expected.txt:
* LayoutTests/platform/ios-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* LayoutTests/platform/mac-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt: Removed.
* LayoutTests/platform/mac/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::attachmentBackgroundIdentifier):
(WebCore::HTMLAttachmentElement::ensureWideLayoutShadowTree):
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-container):
(div#attachment-background):

Canonical link: <a href="https://commits.webkit.org/273642@main">https://commits.webkit.org/273642@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fae18b0f38e351f28ab34163e7063afc73b34af8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36193 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15145 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38399 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38918 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32540 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37418 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17584 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12172 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31219 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36750 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12785 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32102 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11208 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11238 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32285 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40164 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32850 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32657 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37170 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11434 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9318 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35261 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13158 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8219 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11906 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12278 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/RenderThemeMac.mm:
```diff
@@ -602,9 +617,13 @@ static Color activeButtonTextColor()
+            if (auto nsColor = wtfObjCMsgSend<NSColor *>([NSColor class], selector)) {
+                auto color = semanticColorFromNSColor(nsColor);
+                if (makeOpaque)
+                    color = blendSourceOver(localAppearance.usingDarkAppearance() ? Color::black : Color::white, color).semanticColor();
```
You should blend over the system color named Canvas if you can.

squelart replied:
Thanks for the tip, I'll give it a try.

Note that I've copied what was on the iOS side: https://searchfox.org/wubkat/rev/1dfbd6f7279ea1201b5afb9719b4163c8fab1566/Source/WebCore/rendering/RenderThemeIOS.mm#1136
Should I try to change that one to use the "Canvas" color as well? (Maybe in a separate bug?)

nt1m commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -34,13 +34,13 @@ div#attachment-container {
     gap: 4px;
     --information-margin-end: 24px;
     width: 266px;
-    background-color: -apple-system-tertiary-fill;
```
These 2 colors were private fwiw:
https://searchfox.org/wubkat/rev/f57f00ab6a22c7b846e9dda0fa296da77aa48e23/Source/WebCore/css/parser/CSSParserIdioms.cpp#42-47

I think they were added for attachment. Can we replace them altogether? Can we make the new ones private too?

squelart replied:
These colors were chosen by our designers, and were even added to AppKit at about the same time. So I doubt we can replace them.
Or am I missing something, what did you have in mind?

But if we keep them, yes I should make the new derived "opaque" ones private as well.

nt1m replied:
Can we make the new opaque ones private as well?

Btw, can you file a bug about color-mix() not working dynamically with system colors? Ideally we would use it here

squelart replied:
I've made the new opaque colors private.

About `color-mix()`: I'm not sure anymore that it was incorrect, I'll have another look before filing a bug.
But in any case, after experimenting I prefer using `blendSourceOver` here, to ensure a fully opaque background which `color-mix()` wouldn't give us, and to be consistent with what iOS was already doing.

nt1m replied:
Why does color-mix() not give an opaque color? It should if you use it correctly

squelart replied:
I did something like `color-mix(in srgb, quaternary, canvas x%)` (with x at 50% or more), which I understood as computing a weighted sum.
Knowing that quaternary.alpha=0.05 and canvas.alpha=1, how could I possibly get alpha==1 in the end when x<100%?
Is there a way to achieve something like `blendSourceOver` with `color-mix()`?

nt1m commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -27,28 +27,40 @@ div#attachment-container {
+    background-color: -apple-system-tertiary-fill;
+#endif
+    width: 100%;
+    height: 100%;
```
You don't need these 2 lines with display: flex

squelart replied:
If I don't have them, the attachment shrinks to fit the content, which we don't want.

nt1m replied:
Are you sure you need both? I'm pretty sure only one of the two are needed

nt1m replied:
I think you can drop the height

nt1m commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -27,28 +27,40 @@ div#attachment-container {
+    display: flex;
+}
+
+div#attachment-background {
```
Can you add a comment describing why we need this? (e.g. because the system color is semi-transparent)

squelart committed:
Attachment: Use opaque background color to ensure readability
https://bugs.webkit.org/show_bug.cgi?id=267253
rdar://117749005

Reviewed by Tim Nguyen.

The system colors used for the attachment background are very transparent,
which could render the attachment contents less readable if whatever is
under the attachment is of a similar color as the attachment text.
This issue can be prevented by making the background fully opaque.
The original transparent color has been moved to a new child div.

* LayoutTests/fast/attachment/mac/wide-attachment-image-controls-basic-expected.txt:
* LayoutTests/platform/ios-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* LayoutTests/platform/mac-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt: Removed.
* LayoutTests/platform/mac/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::attachmentBackgroundIdentifier):
(WebCore::HTMLAttachmentElement::ensureWideLayoutShadowTree):
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-container):
(div#attachment-background):

Canonical link: https://commits.webkit.org/273642@main
